### DISCUSSION TITLE
Ottavas and cross-staff beams

### DISCUSSION THREAD

**Discussion by mscuthbert** [w3c/mnx#346] [General]
One of the things that seems like a major goal of anyone connected w/ MNX is to avoid having backward and forward entities like the MusicXML tag.  Ran into another unforeseen consequence recently of music21's way of writing MusicXML voices -- all notes from one voice first, backward to the beginning of the measure, all notes from the other second.  What to do for Ottava signs (which E. Gould does confirm always apply to all voices on a staff) that end mid-measure -- the notes must be written out in musicxml in an order that shows which notes in both parts are covered within the ottava; given other multi-voice spanner-type symbols (pedal marks, etc.), it means that for perfect preservation of endings of, each note in musicxml should be written from left to right in order of performance, and voices should be separated not written together.  (Discussion at music21 issue: https://github.com/cuthbertLab/music21/issues/1718)  I will make a note of this on the MusicXML repo.

It seems most logical that 8va-ottavas etc. should go at a higher level than the voice in MNX, and then specify their start and stop according to position.  That solves 99% of problems.  But then what to do about cross-staff beams within ottava signs?  (from Gould Behind Bars, p. 325)

<img width="274" alt="Screenshot 2024-06-25 at 13 23 40" src="https://github.com/w3c/mnx/assets/3521479/020b15af-c3d3-41e9-b699-1a5bcfaf3391">

How to specify that an ottava applies to notes within a voice only when they appear on a certain staff?  Perhaps we will need to have a position+staff-number specification element?  But then things like piano pedals apply to all staves, so that needs a way of specifying (well, generally it does--though you could imagine a piano part with a 3rd staff to play on a synthesizer or a la [Hiromi](https://youtu.be/pnISpahN2dM?si=RWocZWl3hAvy6VwR&t=1310), where the pedal only applies to some staves)

I don't think we've had a discussion of the positioning of spanners like this, and I'd really like to get the weird cases out first, because a solution that works for the weird is usually the best solution for the normal as well.

**Comment by lemzwerg** [comment]
Perhaps an additional `merge-with` attribute will do that specifies the *optical* appearance of ottava lines? I can imagine that the main representation in MNX is rather this:

![image](https://github.com/w3c/mnx/assets/1630245/cd24a524-d0de-4cd5-a2a2-7f48f83337b3)

Assuming IDs `ottava-1`, `ottava-2`, and `ottava-3`, the second and third element would also have `merge-with="ottava-1"`.

**Reply by adrianholovaty** [reply]
Though it's possible to encode a document this way (with three separate octave-shift objects), I'd strongly recommend encoding it a single octave-shift. [See my comment here](https://github.com/w3c/mnx/discussions/346#discussioncomment-9891410).

**Comment by adrianholovaty** [comment]
We're mostly there with the current MNX spec. The [octave shift object](https://w3c.github.io/mnx/docs/mnx-reference/objects/octave-shift/) lives within a sequence, directly before the first note it affects, and it has an `"end"` value that describes the rhythmic position of the last note it affects.

Here's an example document: https://w3c.github.io/mnx/docs/mnx-reference/examples/octave-shifts-8va/

To handle the case of ottavas in cross-staff music, I think there are two missing pieces:

* We haven't yet added a way to encode cross-staff notes. I'm planning to do this in the coming days (it'll be a simple `"staff"` key on note objects).
* We haven't explicitly documented the behavior of octave shifts for cross-staff notes. It might be good enough to simply update the octave shift docs to say "An octave shift only applies to the notes in the current staff." We could also expand it with an optional key that specifies "this octave shift applies to both staves."

**Reply by lemzwerg** [reply]
Sounds good, except this:

>  We could also expand it with an optional key that specifies "this octave shift applies to both staves."

IMHO, an ottava line is strictly bound to a single staff – or rather, in same (not ideal) cases, to the top-most voice in a staff.  Do you have a real-world example of good typography where an ottava line applies to two staves?

**Reply by adrianholovaty** [reply]
I agree, and I don't have any examples for you. I was merely pointing out that it's an option to extend the spec if needed (in other words, the structure of the encoding is extensible in the future in an elegant way).

**Reply by lemzwerg** [reply]
OK, good to know.

BTW, the extension should be 'this ottava line belongs only to the top-most (or bottom-most) voice'.

Here is an example from Debussy, *Préludes* 2ᵉ Livre, *... La terrasse des audiences du clair de lune*.

![image](https://github.com/w3c/mnx/assets/1630245/b617276a-918c-484f-aff5-f4551f0f3ddb)

**Reply by adrianholovaty** [reply]
Thanks for the example. Octave shifts are already encoded within a specific voice (a "sequence" in MNX parlance), so perhaps it could be "this applies only to this sequence" versus "this applies to all sequences in this staff." I'd like to avoid the ideas of top and bottom voices if possible, as that's a bit brittle.
